### PR TITLE
resource/db_instance: add restore to point in time support

### LINKED ISSUE #5286: point in time restore support for aws rds instances
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Today terraform aws provider support managing rds instances through aws_db_instace resource.
https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_db_instance.go
This resource also supports restoring from a snapshot if snapshot_identifier is passed and also supports restore db instance from s3.
We don't seem to have support for restoring to a specific point in time. This maps to below AWS API
https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_RestoreDBInstanceToPointInTime.html

We are trying to reduce the restore times when an incident occurs that requires db instance restore so that we can meet our RTO (and RPO) and quickly bring the production back to business. 
Having an automated restore procedure in place is very critical. In our case we are using aws rds postgres and also have plans to move to aurora in future.
Since we would like Terraform to manage the existing and newly created resources (from restore), we are looking forward for the restore to a point in time support in terraform (for rds instances as of now).

Is this in the road map? Can some one share details on timeline or what is needed to get this feature support? I am new to the go programming but am myself looking at the aws provider code base to see what can be done. Any help is appreciated.

### New or Affected Resource(s)

* aws_db_instance

### References
I see some discussion in google group here https://groups.google.com/d/topic/terraform-tool/vDUw9YmqtN4/discussion and a suggestion from Paul Hinze.
AWS API
https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_RestoreDBInstanceToPointInTime.html

### PR DESCRIPTION
**Pull request by anGie44**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #5286
Preceded by #7031 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/db_instance: add `restore_to_point_in_time` argument
resource/db_instance: add `latest_restorable_time` attribute
```

Output from acceptance testing:

```
--- PASS: TestAccAWSDBInstance_RestoreToPointInTime_SourceIdentifier (1469.96s)
--- PASS: TestAccAWSDBInstance_RestoreToPointInTime_SourceResourceID (1275.53s)

--- PASS: TestAccAWSDBInstance_kmsKey (430.59s) (previously panicking but fixed w/ commit 81ca42c14)

other test results in progress...
```

### CODE REVIEW

File: aws/validators_test.go [hunk 1]
```diff
@@ -3220,3 +3220,29 @@ func TestValidateServiceDiscoveryNamespaceName(t *testing.T) {
...
+		}
+	}
+
+	for _, f := range invalidT {
+		_, errors := validateUTCTimestamp(f, "invalid UTC timestamp")
+		if len(errors) == 0 {
+			t.Fatalf("Expected the time %q to fail validation", f)
+		}
+	}
+}
```
Comment (anGie44) [inline aws/validators_test.go:3248]: copied over from https://github.com/terraform-providers/terraform-provider-aws/pull/7031 to re-use here though under a more generic func name

File: aws/resource_aws_db_instance.go [hunk 2]
```diff
@@ -273,6 +278,41 @@ func resourceAwsDbInstance() *schema.Resource {
...
+				},
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"restore_time": {
+							Type:          schema.TypeString,
+							Optional:      true,
+							ValidateFunc:  validateUTCTimestamp,
+							ConflictsWith: []string{"restore_to_point_in_time.0.use_latest_restorable_time"},
+						},
+						"source_db_instance_identifier": {
```
Comment (YakDriver) [inline aws/resource_aws_db_instance.go:299]: Nit: I like lines between.
```suggestion

						"source_db_instance_identifier": {
```

File: aws/resource_aws_db_instance.go [hunk 3]
```diff
@@ -273,6 +278,41 @@ func resourceAwsDbInstance() *schema.Resource {
...
+							Type:     schema.TypeString,
+							Optional: true,
+						},
```
Comment (YakDriver) [inline aws/resource_aws_db_instance.go:302]:
```suggestion
						},

```

File: aws/resource_aws_db_instance.go [hunk 4]
```diff
@@ -273,6 +278,41 @@ func resourceAwsDbInstance() *schema.Resource {
...
+						"source_dbi_resource_id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
```
Comment (YakDriver) [inline aws/resource_aws_db_instance.go:306]:
```suggestion
						},

```

File: aws/resource_aws_db_instance.go [hunk 5]
```diff
@@ -1816,3 +1928,31 @@ var resourceAwsDbInstanceUpdatePendingStates = []string{
...
+
+func expandRestoreToPointInTime(l []interface{}) *rds.RestoreDBInstanceToPointInTimeInput {
+	if len(l) == 0 || l[0] == nil {
+		return nil
+	}
+	tfMap, ok := l[0].(map[string]interface{})
+	if !ok {
+		return nil
+	}
+	input := &rds.RestoreDBInstanceToPointInTimeInput{}
```
Comment (YakDriver) [inline aws/resource_aws_db_instance.go:1940]: Nit: Between groups of statements, I like a line.
```suggestion

	input := &rds.RestoreDBInstanceToPointInTimeInput{}
```

File: aws/resource_aws_db_instance.go [hunk 6]
```diff
@@ -1816,3 +1928,31 @@ var resourceAwsDbInstanceUpdatePendingStates = []string{
 	"storage-full",
 	"upgrading",
 }
+
+func expandRestoreToPointInTime(l []interface{}) *rds.RestoreDBInstanceToPointInTimeInput {
+	if len(l) == 0 || l[0] == nil {
+		return nil
+	}
+	tfMap, ok := l[0].(map[string]interface{})
```
Comment (YakDriver) [inline aws/resource_aws_db_instance.go:1936]:
```suggestion

	tfMap, ok := l[0].(map[string]interface{})
```

File: aws/resource_aws_db_instance.go [hunk 7]
```diff
@@ -1816,3 +1928,31 @@ var resourceAwsDbInstanceUpdatePendingStates = []string{
...
+		return nil
+	}
+	input := &rds.RestoreDBInstanceToPointInTimeInput{}
+	if v, ok := tfMap["restore_time"].(string); ok && v != "" {
+		parsedTime, err := time.Parse(time.RFC3339, v)
+		if err == nil {
+			input.RestoreTime = aws.Time(parsedTime)
+		}
+	}
+	if v, ok := tfMap["source_db_instance_identifier"].(string); ok && v != "" {
```
Comment (YakDriver) [inline aws/resource_aws_db_instance.go:1947]:
```suggestion

	if v, ok := tfMap["source_db_instance_identifier"].(string); ok && v != "" {
```

File: aws/resource_aws_db_instance_test.go [hunk 8]
```diff
@@ -3552,6 +3622,55 @@ resource "aws_db_instance" "test" {
...
 
+const testAccAWSDBInstanceBaseConfig = `
+resource "aws_db_instance" "test" {
+  allocated_storage       = 10
+  backup_retention_period = 1
+  engine                  = data.aws_rds_orderable_db_instance.test.engine
+  engine_version          = data.aws_rds_orderable_db_instance.test.engine_version
+  instance_class          = data.aws_rds_orderable_db_instance.test.instance_class
+  name                    = "baz"
+  parameter_group_name    = "default.mysql5.6"
```
Comment (YakDriver) [inline aws/resource_aws_db_instance_test.go:3633]: Ugh. This is my doing but we should be using a `aws_rds_engine_version` in this config... Different PR.

File: aws/validators.go [hunk 9]
```diff
@@ -1956,14 +1956,16 @@ func validateIoTTopicRuleCloudWatchAlarmStateValue(v interface{}, s string) ([]s
 	return nil, []error{fmt.Errorf("State must be one of OK, ALARM, or INSUFFICIENT_DATA")}
 }
 
-func validateIoTTopicRuleCloudWatchMetricTimestamp(v interface{}, s string) ([]string, []error) {
+// validateUTCTimestamp validates a string in UTC Format required by APIs including:
+// https://docs.aws.amazon.com/iot/latest/apireference/API_CloudwatchMetricAction.html
+// https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_RestoreDBInstanceToPointInTime.html
+func validateUTCTimestamp(v interface{}, k string) (ws []string, errors []error) {
```
Comment (YakDriver) [inline aws/validators.go:1962]: Nice

File: website/docs/r/db_instance.html.markdown [hunk 10]
```diff
@@ -198,15 +203,27 @@ for more information.
...
+### Restore To Point In Time
+
+-> **Note:** You can restore to any point in time before the source DB instance's `latest_restorable_time` or a point up to the number of days specified in the source DB instance's `backup_retention_period`.
+For more information, please refer to the [Developer Guide](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_PIT.html).
+This setting does not apply to `aurora-mysql` or `aurora-postgresql` DB engines. For Aurora, refer to the [`aws_rds_cluster` resource documentation](https://www.terraform.io/docs/providers/aws/r/rds_cluster.html#restore_in_time).
+
+The `restore_to_point_in_time` block supports the following arguments:
+
+* `restore_time` - (Optional) The date and time to restore from. Value must be a time in Universal Coordinated Time (UTC) format and must be before the latest restorable time for the DB instance. Cannot be specified with `use_latest_restorable_time`.
+
```
Comment (YakDriver) [inline website/docs/r/db_instance.html.markdown:220]: I don't know the rule on this but sometimes there is a blank line in this situation and other times not...

**Review by YakDriver — APPROVED** [review]
A few nits but looks great!

GovCloud:
```
--- PASS: TestAccAWSDBInstance_MinorVersion (429.15s)
--- PASS: TestAccAWSDBInstance_basic (432.28s)
--- PASS: TestAccAWSDBInstance_AllowMajorVersionUpgrade (447.87s)
--- PASS: TestAccAWSDBInstance_namePrefix (449.64s)
--- PASS: TestAccAWSDBInstance_IsAlreadyBeingDeleted (459.03s)
--- PASS: TestAccAWSDBInstance_DbSubnetGroupName_VpcSecurityGroupIds (474.81s)
--- PASS: TestAccAWSDBInstance_RestoreToPointInTime_SourceResourceID (1305.06s)
--- PASS: TestAccAWSDBInstance_RestoreToPointInTime_SourceIdentifier (1388.71s)
```

File: website/docs/r/db_instance.html.markdown (hunk 10 above)
Comment (anGie44) [reply website/docs/r/db_instance.html.markdown:220]: hmm true..i'll follow the convention in this file -- seems there aren't blank lines between arguments

File: aws/resource_aws_db_instance_test.go (hunk 8 above)
Comment (anGie44) [reply aws/resource_aws_db_instance_test.go]: @YakDriver I gave it a go with the new data source in the latest commit -- looks ok?

File: aws/validators_test.go [hunk 11]
```diff
@@ -3220,3 +3220,29 @@ func TestValidateServiceDiscoveryNamespaceName(t *testing.T) {
...
+	invalidT := []string{
+		"2015-03-07 23:45:00",
+		"27-03-2019 23:45:00",
+		"Mon, 02 Jan 2006 15:04:05 -0700",
+	}
+
+	for _, f := range validT {
+		_, errors := validateUTCTimestamp(f, "UTC timestamp")
+		if len(errors) > 0 {
+			t.Fatalf("Expected the time %q to be in valid format, got error %q", f, errors)
```
Comment (YakDriver) [inline aws/validators_test.go:3238]: More nits:
```suggestion
			t.Fatalf("expected the time %q to be in valid format, got error %q", f, errors)
```

File: aws/validators_test.go [hunk 12]
```diff
@@ -3220,3 +3220,29 @@ func TestValidateServiceDiscoveryNamespaceName(t *testing.T) {
...
+		}
+	}
+
+	for _, f := range invalidT {
+		_, errors := validateUTCTimestamp(f, "invalid UTC timestamp")
+		if len(errors) == 0 {
+			t.Fatalf("Expected the time %q to fail validation", f)
```
Comment (YakDriver) [inline aws/validators_test.go:3245]:
```suggestion
			t.Fatalf("expected the time %q to fail validation", f)
```

**Pull request merged**